Dage 21/ Vespa Search Engine + Unit test + Integration with simple scripts

Intrinsical-AI opened the pull request:
### Ticket
 [**DAGE-21**](https://sease.atlassian.net/jira/software/projects/DAGE/boards/84?selectedIssue=DAGE-21)

- Implemented the Vespa Search Engine 
- Implemented Mocks + Unit Testing following the previous Solr pattern
- Added sample config file (good_config_vespa)
- Added vespa-init/ with:
  -- Makefile -> To handle flows, in this case with make ci, we're supposed to setup + launch + lightweight test on the Docker Instance (not added pytest-docker yet)
  -- docker compose -> Simple, I had problems configuring it
  -- bash script
  -- python testing script 
  -- /dataset -> folder to copy to docker container with files
  -- /app -> Vespa required - with schemas and service .xml files
  
***IMPORTANT:*** check the vespa-init/CHEETSHEET.md to get an overview of the used commands

Intrinsical-AI committed:
Added vespa-init folder to the integration tests directory, mimicing solr-init previous approach. Added compose + docker + vespa_init.sh + test_vespa.py + cheetsheet + Makefile (make ci for git / fast testing). Added vespa setup + schemas into vespa-init/app (services, news.d)..

Intrinsical-AI committed:
Merge branch 'dataset-generator' into DAGE-21/Vespa-Search-Integration

Intrinsical-AI committed:
Synced with main branch

Intrinsical-AI committed:
Improved styling, fixed minors

Intrinsical-AI committed:
removed intruder file

Intrinsical-AI committed:
clearifying notes to setup / init and testing

Intrinsical-AI committed:
clarifying notes to setup / init and testing

Intrinsical-AI committed:
improved Cheetsheet and Makefile

Intrinsical-AI committed:
Added unit tests for Vespa Search Engine following the Solr pattern - Mocks and config included

Intrinsical-AI committed:
parametrize different docs to validate search with filters using the template

Intrinsical-AI committed:
Polish docstrings

Intrinsical-AI commented:
Note: still pending "launch" standarization between different engines - TBD

nseidan commented on rre-dataset-generator/tests/integration/vespa-init/vespa-init.sh:
```diff
@@ -0,0 +1,70 @@
+
+docker exec -it vespa bash -c "vespa query \"select * from news where true\" language=en-US"
+# EXPECTED:
+# {
```
I suspect we don't need these commented lines. Any strong reasoning?

Intrinsical-AI replied:
You're totally right, inthis case, was even breaking the execution. Removed in the next commit

Intrinsical-AI committed:
Added changes, based on the reviews made on the rest of the engines - aligned structure and implementation

Intrinsical-AI committed:
Aligned structure, added exhaustive tests (unit), Makefile for smooth handling, singe command running, vespa search engine with strict checks on payloads, used pytests with fixtures and parametrization

Intrinsical-AI committed:
Aligned structure, added exhaustive tests (unit), Makefile for smooth handling, singe command running, vespa search engine with strict checks on payloads, used pytests with fixtures and parametrization

Intrinsical-AI commented:
Mini Update:

* Added full **Vespa** support: new `VespaSearchEngine`, safe YQL queries, field validation, and robust parsing.
* New **Docker Compose + Makefile** setup for Vespa integration and schema/data bootstrapping.
* Added **unit and integration tests** with mocks and config compatibility.
* Updated **README** with Vespa usage instructions, and added a **CHEETSHEET** file 
* Improved docstrings - Following *Google Stylesheet for Docstrings*


CHECK https://sease.atlassian.net/browse/DAGE-21?focusedCommentId=16702 for in depth details

Intrinsical-AI committed:
upgraded tests - prune + minors

nicolo-rinaldi commented on rre-dataset-generator/src/search_engine/vespa_search_engine.py:
```diff
@@ -0,0 +1,289 @@
+        # (Vespa uses the fields declared in the YQL).
+        payload = {
+            "yql": yql,
+            "hits": min(10, MAX_HITS),
```
i think this is a bit a constraint. in solr at least, I assumed that the hits param was inside the template or, if not there, i sticked to 10

Intrinsical-AI replied:
Noted along the clean_text comments -> In general: harder and extensive validations would be nice to the Config. 

A good option in the future would be to implement "pagination" /batching but thats complex and out the scope of this stage, I belive.

 So yes, I agree. Back to the "default-without-constraints", it's the better approach here and now.

Intrinsical-AI replied:
**[UPDATED]**
"hits": max(0, int(doc_number))

Intrinsical-AI committed:
upgraded tests - prune + minors

Intrinsical-AI committed:
styles: removing unused imports

Intrinsical-AI commented on rre-dataset-generator/tests/integration/vespa-init/app/services.xml:
```diff
@@ -0,0 +1,75 @@
+        <nodes>
+            <node hostalias="node1" distribution-key="0" />
+        </nodes>
+        <engine>
```
This part <engine>, is not needed, but added to fulfill local constraints

Intrinsical-AI commented on rre-dataset-generator/tests/integration/vespa-init/data/dataset.json:
```diff
@@ -0,0 +1,8 @@
+{
+    "put": "id:news:news::1",
```
This is for testing purposes. As the Vespa expected format differs, We would need a script to pass from the used json in the rest of the engines, to this format.

A good approach migh be adding a special general parser into the VespaEngine

nicolo-rinaldi replied:
Is this way of adding documents creating a duplicate of the id?

Intrinsical-AI replied:
***Documented Here***

`id:news:news::1 `
corresponds to:
`id:<namespace>:<document-type>::<user-specific-id> `
 [Vespa-Docs](https://docs.vespa.ai/en/reference/document-v1-api-reference.html)

> So this creates OR **replaces** / overrides an object type "news", in the namespace "news", with ID "1".

***NOTE***: Vespa accepts every namespace (alpha-num string) and treat it as a logical split. The namespace it's not required to be defined previously!

---

Also: Vespa exposes a REST API (different from other engines). Currently we're relying on the vespa cli, and using  `vespa feed data/*.json` [ vespa-init.sh ] inside the container to load the data. We could 'craft' a lower level method to interact with the API, but seems simpler to implement a parser inside the vespa engine with parametrizations + defaults, when the moment comes

Intrinsical-AI committed:
Comment clarification

Intrinsical-AI committed:
Polished comments / explainations on the code

Intrinsical-AI committed:
updated log level on the vespa tests

Intrinsical-AI committed:
Added tests Docstrings - adapted for devs. Minor refinements.

Intrinsical-AI committed:
removed unused function

Intrinsical-AI committed:
minors

nicolo-rinaldi commented on rre-dataset-generator/src/search_engine/vespa_search_engine.py:
```diff
@@ -0,0 +1,337 @@
+
+        # WARNING: clean_text it's unly used to clean html tags and unescape html entities
+        # > it's not used to clean the text of the document
+        # TODO: proposal - refactor clean_text to be consistent with the name, and intended functionality
```
feel free to do it

Intrinsical-AI replied:
When I finish this PR I'll improve it.

 I don't want to impact yet common files while we have 3 different PRs pending to approve (Naz, yours, mine). Makes sense to maintain the comment so we don't forget / we notice

nseidan replied:
it's better to create a jira issue for that, but not write a comment here. I think It doesnt look good like this comment in the codebase;

nicolo-rinaldi commented on rre-dataset-generator/tests/integration/docker-compose.vespa.yaml:
is the healthchek part used instead of `*-init.sh` file in the other search engines?

Intrinsical-AI replied:
Oh, got you wrong at first!

Yes, I could add it to the .sh script directly.
Makes more sense

Intrinsical-AI replied:
**[UPDATED]**

nicolo-rinaldi commented on rre-dataset-generator/tests/integration/vespa-init/CHEETSHEET.md:
````diff
@@ -0,0 +1,99 @@
+Manual checks:
+
+```bash
+python -m pytest tests/integration/vespa-init/vespa_test_manual_integration.py
````
remember that we are using uv. Up to now, the bash script is 
```
uv run pytest <optional-folder-containing-tests>
```

Intrinsical-AI replied:
You're right! This is outdated, should not be here since I removed the manual integration test. Will be updated/fixed in next commit

Anw: uv is very good for installing packages. For dev I rather use a mix of bash commands and/or modules (ej: source, python -m venv, directly pytests). The point is: there's a lot of alternatives, and uv it's not the standard  for all of them (neither the absolute best), usually dependes on the context.

 For the docs, 100% agree should be unified - uv seems a good option

nicolo-rinaldi commented on rre-dataset-generator/tests/integration/vespa-init/app/schemas/news.sd:
seems good to me

Intrinsical-AI replied:
This is copied directly from the Vespa tutorial (service.xml also), the news.sd was intially but then refactored to adapt the structure of the test dataset

nicolo-rinaldi commented on rre-dataset-generator/tests/integration/vespa-init/Dockerfile:
```diff
@@ -0,0 +1,9 @@
+FROM python:3.10-slim
```
why do we want to add python? maybe I already asked this question, but it seems strange to me. This is a python-based project and everything we need I think should be in the the environment

Intrinsical-AI replied:
I think I needed for executing the manual integration test ( it was executed **from** the container) uploaded in previous commits but deleted in the last one. Will re-check if any script executed in the container relies on Python. 

Anw: Docker image does not copy any current env configuration by default. Everything you're gonna need to *execute inside the container* have to be explicit about the installation in the Dockerfile.

Intrinsical-AI replied:
Man, just spoted it. It's just under this comment. pip requires Python hahahaha

 Yes, was for the manual tests. Still, as exaplained this is only being installed in the cointainer, and will need it / be usefull for the integrations. I'd keep it. Lmk if you're sure you prefer it out

nicolo-rinaldi commented on rre-dataset-generator/tests/integration/vespa-init/docker-compose.yml:
I see 2 docker compose very similar. Why we need 2? I see that one is calling `vespa-init.sh` and the other is not. Is this because one is supposed to index docs and the other is not?

Intrinsical-AI replied:
this was supposed to be deleted. Update in next commit

nicolo-rinaldi commented on rre-dataset-generator/tests/mocks/vespa.py:
```diff
@@ -0,0 +1,32 @@
+    The payload structure mirrors the real Vespa JSON structure that the
+    VespaSearchEngine expects: {"root": {"children": [...]}}
+    """
+    def __init__(self, docs, status_code: int = 200):
```
typing is missing

Intrinsical-AI replied:
You got me, but I see you added it. I will copy you :+1:

nicolo-rinaldi commented on rre-dataset-generator/tests/integration/vespa-init/vespa-init.sh:
```diff
@@ -0,0 +1,36 @@
+echo "Running test query …"
+vespa query --target $HTTP_URL "select * from news where true" language=en-US
+
+echo "Vespa initialization complete."
```
new line

Intrinsical-AI replied:
oka 

:-1: to the red circle

nicolo-rinaldi commented on rre-dataset-generator/tests/unit/test_vespa_search_engine.py:
```diff
@@ -0,0 +1,352 @@
+
+from tests.mocks.vespa import MockResponseVespaSearch
+
+configure_logging()
```
I'd rather set this to debug

Intrinsical-AI replied:
Ok, you're right! 

I'm explictly setting the value to expect in each test, but adding level debug might add useful information while testing

Good spot

nicolo-rinaldi commented on rre-dataset-generator/tests/unit/test_vespa_search_engine.py:
```diff
@@ -0,0 +1,352 @@
+            },
+        },
+        {
+            "id": "id:news:news::2",
```
why here there is a second document

Intrinsical-AI replied:
Check @pytest.mark.parametrize 

This is getting executed 2 times - one for each test defined

**UPDATE**

Re-checking: you might be question why not adding a general doc_mock with pytests fixtures.

 If that: 
 - It's intended: think this is as a different case than the "Config" fixture, as we're trying to "break" each function with strange and rare cases on each step / internal component.
- Having multiple tests with the same mock, would be mis-aligned with the test intentions
 
 Hope it makes sense! > I'm just learning how much important this is for testing

Intrinsical-AI replied:
Thanks for the feedback in general btw @nicolo-rinaldi :+1: 

Hitting you on Slack as soon as I upload the patch / fixes

Intrinsical-AI commented on rre-dataset-generator/src/search_engine/vespa_search_engine.py:
```diff
@@ -0,0 +1,337 @@
+            return "true"
+
+        clauses: List[str] = []
+        for f in filters:
```
**AUTOFEEDBACK**

(I know Daniele will catch this anyway, so posting it here.)

Example:
```
filters = [
    {"field_1": ["a", "b"]},
    {"field_1": ["c"]}
]
```

This generates:

```
(field_1 contains "a" OR field_1 contains "b") AND (field_1 contains "c")
```
Syntax valid **BUT**.. logic should go from:

CURRENT
(field_1 contains "a" OR field_1 contains "b") AND (field_1 contains "c") 
> Silent logical failure - eg: scalars ("scores") only allows 1 value at time, so the current impl. will return empty results always for scalar values, even we intended to express other logical statement that has hits.


TO (***group by field and then apply filters***):
(field_1 contains "a" OR field_1 contains "b" OR field_1 contains "c")
> Valid and coherent - including scalar (only allows one value at each time so the current impl. will return empty results always)

Intrinsical-AI replied:
**[No action needed]**
 
This should be solved now

nseidan commented on rre-dataset-generator/tests/integration/vespa-init/app/services.xml:
```diff
@@ -0,0 +1,75 @@
+
+
+
+    
```
why push app/schemas/vespa-init.sh if it is empty?

Intrinsical-AI replied:
Execution artifact - removing it. Thanks

nseidan commented on rre-dataset-generator/tests/integration/vespa-init/vespa-init.sh:
```diff
@@ -0,0 +1,36 @@
+# Feed the sample data
+echo "Feeding data …"
+vespa feed --target $HTTP_URL /data/*.json
+sleep 5
```
hardcoded sleep here? how about checking docs stats/counts if they have been indexed ? 
Also, if the server restarts , this will again index the dataset,but the dataset is already there. We want to index dataset only if we haven't indexed;

Intrinsical-AI replied:
You're right to better checking the docs. In this case; I'm not so sure that docker _always_ keeps the instance with the index, but you're right.

Will be updating it, thanks!

nseidan commented on rre-dataset-generator/tests/integration/vespa-init/vespa-init.sh:
```diff
@@ -0,0 +1,36 @@
+
+# Deploy the application package
+echo "Deploying Vespa application (config server $CONFIG_URL)…"
+vespa deploy --wait 300 --target $CONFIG_URL /app
```
in the case of `vespa deploy` failure (e.g. config issue) the script continues without actually deploying.

Intrinsical-AI replied:
I think it didn't!
The makefile included a default check. 
Nicolo also pointed that, so I migrated it to the vespa-init.sh in the last commit! 
Now should be clearer.

 PS: Anw: I suggest checking the Makefile, in general are very useful and flexible

nseidan commented on rre-dataset-generator/tests/integration/docker-compose.vespa.yaml:
```diff
@@ -0,0 +1,18 @@
+    ports:
+      - "8080:8080"
+      - "19071:19071"
+    volumes:
```
for solr engine, we didn't have to handle volumes, with the 'solr-init' trick. We could do similar here as well.

Intrinsical-AI replied:
Hmmm you might be right, but I'm not so sure there's a big difference, volumes could be usefull in the future. Also: running and testing Vespa with Compose was a bit of a headache. 

Since I don't see direct impact now removing the volumes, I'd vote to keep it as this for the moment, since it's tested and working. Maybe you spot a critical difference. If so, let me know and I'll update, but tbh I'd keep as this for now.

nseidan commented on rre-dataset-generator/tests/integration/docker-compose.vespa.yaml:
```diff
@@ -0,0 +1,18 @@
+      test: ["CMD", "curl", "-s", "--head", "http://localhost:19071/ApplicationStatus"]
+      interval: 1s
+      timeout: 10s
+      retries: 60
```
why docker-compose.yml is removed?

Intrinsical-AI replied:
I don't get this? We have a custom docker.compose per engine, right? 
I think this is the correct structure, aligned with dataset-generation, let me know if I'm missing something

nseidan commented on rre-dataset-generator/tests/integration/vespa-init/docker-compose.yml:
```diff
@@ -0,0 +1,18 @@
+services:
+  vespa:
+    image: vespaengine/vespa
```
better to specify the version as the latest one might not be always stable.

Intrinsical-AI replied:
This is very smart, you're right. I will think about it for next time.

 For now, since this is working (executed + checked), and this is only for testing as you proposed in the OpenSearch Pull, we might keep like this, and update if needed as the simplest option? What you think?

nseidan replied:
is there any problem if you specify a version and pull the image ?

Intrinsical-AI committed:
Polish + Refinements based on Nico`s

Intrinsical-AI committed:
Updated with dataset-generator branch. Updated README with Vespa Standalone setup

Intrinsical-AI committed:
Updates requested by Nicolo and Naz feedback - (cleanup:) removed files (docker-compose + app/vespa-init.sh) - added debug to unit/test_vespa and healthchecks to vespa-init.sh

Intrinsical-AI committed:
Update test naming and adapt to the new logic for the hit number on the query_generation payload (vespa_search_engine)

Intrinsical-AI committed:
Updated vespa-init.sh script following Naz feedback to avoid hardcoded sleeps (and implementing healthcheks inside the script) - pending to local test

nseidan commented on rre-dataset-generator/tests/integration/docker-compose.yml:
```diff
@@ -1,15 +0,0 @@
-services:
-  solr:
-    image: solr:9.8.1
-    container_name: solr_test
```
don't remove this, it is for solr.

Intrinsical-AI committed:
Polished with Naz Feedback: added last docker compose for Solr, and aligned README with the ElasticSearch PR

nseidan commented on rre-dataset-generator/tests/integration/Makefile:
```diff
@@ -0,0 +1,36 @@
+# Run the initialization script inside the Vespa container
+vespa-init:
+	@echo "Waiting for Vespa admin endpoint to respond..."
+	@timeout 300 bash -c 'until curl -s --head http://localhost:19071/ApplicationStatus | grep "200 OK" >/dev/null; do sleep 1; done'
```
in `vespa-init.sh`, you perform health check again, so it is redundant.

Intrinsical-AI committed:
Polished with Naz Feedback: added last docker compose for Solr, and aligned README with the ElasticSearch PR

Intrinsical-AI committed:
Polished with Naz Feedback: added last docker compose for Solr, and aligned README with the ElasticSearch PR

Intrinsical-AI committed:
Polished with Naz Feedback: added last docker compose for Solr, and aligned README with the ElasticSearch PR

dantuzi commented on rre-dataset-generator/README.md:
````diff
@@ -126,3 +126,56 @@ This will start 2 services:
+
+##### Running Elasticsearch (Single Node)
+
+Similarly to Solr, to run a local Elasticsearch test environment using docker-compose:
+```bash
+cd tests/integration/
+```
+
+Depending on your Docker version, you may need to use `docker compose` instead of `docker-compose`.
+If you have Docker Compose v1 installed, use:
+
+```bash
+docker-compose -f docker-compose.elasticsearch.yml up --build 
+```
+If you have Docker Compose v2 installed, use:
+```bash
+docker compose -f docker-compose.elasticsearch.yml up --build
+```
+
+This will start 2 services:
+ - `elasticsearch`, available at http://localhost:9200
+ - `elasticsearch-init`, loads documents from elasticsearch-init/data/dataset.jsonl only if Elasticsearch doesn't have 
+any documents in the index.
+
````
How come this is here? That piece of code has not been merged yet. Is it possible that there was an issue with git?

Intrinsical-AI replied:
This is a lapsus. 

I was looking at the latest docs to follow the same structure with the Vespa content, and I added this part manually by accident

Intrinsical-AI commented:
# Update 23-09-2025

> This PR still unreviewed.

 Since the repo and the git history changed a lot during this time, I polished and updated the implementation to align; so **I created a new PR**:
[PR#228](https://github.com/SeaseLtd/rated-ranking-evaluator/pull/228)

Closing this one (deprecated).
